### PR TITLE
[experiment] Configurable mapping of columns to inputs

### PR DESCRIPTION
**Pull request by mikz**
This is a proof of concept to offer ability to configure column to input mapping per form builder.

Some people want to map a column to different input type (#1165, #1143). Some people want a default (#1183). Some would like to have an override (#1187).

proposed in https://github.com/justinfrench/formtastic/issues/1146#issuecomment-203114072

I'd like to get some comments on this approach and verify it solves problems raised in #1146.
I'm not saying I'm happy with this particular API, it is too early for that. But IMO the way is right.

/cc @justinfrench @timoschilling @jclay
## Examples

I have not tried them. But they should work. The idea is to verify the API is ok for those use cases and then actually implement it and have a test coverage for those.
### Adding inet (https://github.com/justinfrench/formtastic/issues/1146#issuecomment-127635091)

``` ruby
string_input = Formtastic::FormBuilder.input_mapping.add_input(:string)
string_input.match_type(:inet)
```
### Mapping Hstore to text (#1165)

``` ruby
text_input = Formtastic::FormBuilder.input_mapping.add_input(:text)
text_input.match_type(:hstore)
```
### Mapping color to string (#1143) (not implemented yet)

``` ruby
string_input = Formtastic::FormBuilder.input_mapping.prepend_input(:string)
string_input.match_column(/color/)
```
### Having global default (#1183) (should change)

```
string_input = Formtastic::FormBuilder.input_mapping.add_input(:string)
string_input.match_type(nil)
```
### Having a name without a type (#1187)

```
password_input = Formtastic::FormBuilder.input_mapping.add_input(:password)
password_input.match_type(nil)
password_input.match_column(/password/)
```
### TODO
- [ ] figure out better way to handle defaults than `Input#match_type(nil)`
- [ ] on form builder inheritance, the input mapping gets copied and possibly mutated

### CODE REVIEW

**Comment by mikz** [comment]
Ping @justinfrench @timoschilling @jclay

No rush. Thaler your time. Would be cool to know if you feel it is right approach or not.

**Comment by justinfrench** [comment]
Hi @mikz, this is a really cool experiment, but I don't _think_ I like it. There's lots to discuss, but perhaps I'll start with two small example.

```
select = input_mapping.add_input(:select)
select.match_type(:integer)
select.match_form { |form, method| form.reflection_for(method) }
select.match_form { |form, method| form.enum_for(method) }

# ...

number = input_mapping.add_input(:number)
number.match_type(:integer)
number.match_type(:float)
number.match_type(:decimal)
```

I have two main observations:

The first is that I can see two inputs being added/registered, both of which `match_type(:integer)`, but it's not clear to me which one has precedence. Is it the first one added? The last one? I agree that the way we currently do it isn't _configurable_, but at least it's much clearer which rules take precedence.

The second is that It's not really clear to me if the `match_*` rules are `AND` or `OR`. I _think_ they're an `OR` — For the Number input you're saying "Register a Number input, use it for Integer or Float or Decimal columns".

That's pretty clear, but for the Select input it gets a bit trickier: "Register a Select input, and use if for Integer columns or when there's a AR reflection or when there's an Enum". Like I said above, we have two different inputs for Integer columns, but the code we currently have in master isn't an `OR`. We have a combination of `AND` and `OR`:

```
when :integer
  return :select    if reflection_for(method)
  return :select    if enum_for(method)
  return :number
```

* Use a select input if the column is an integer AND (there's a reflection OR there's an enum)
* Otherwise, use a number input if the column is an Integer

In short, I think this API makes it far less clear what's happening in a pretty complex decision tree that we've evolved over a long time. I'm not 100% sure what "better" would look like though!

**Comment by mikz** [comment]
@justinfrench awesome feedback! I have no idea how the ideal interface would look like, but maybe we can get there with some experimentation :) Will give it some time and try to come up with another experiment.

**Comment by justinfrench** [comment]
Closing for now due to inactivity
